File: src/model/file.ts
import { v4 as uuid } from 'uuid';

import { SentFileMessage } from 'src/api/types';

import { TGFSFileObject, TGFSFileVersionSerialized } from './message';

export class TGFSFileVersion {
  static EMPTY_FILE = -1;
  static INVALID_FILE_SIZE = -1;

  id: string;
  updatedAt: Date;
  messageId: number;
  size: number;

  toObject(): TGFSFileVersionSerialized {
    return {
      type: 'FV',
      id: this.id,
      updatedAt: this.updatedAt.getTime(),
      messageId: this.messageId,
      size: this.size,
    };
  }

  static empty(): TGFSFileVersion {
    const tgfsFileVersion = new TGFSFileVersion();
    tgfsFileVersion.id = uuid();
    tgfsFileVersion.updatedAt = new Date();
    tgfsFileVersion.setInvalid();
    return tgfsFileVersion;
  }

  static fromSentFileMessage(msg: SentFileMessage): TGFSFileVersion {
    const tgfsFileVersion = new TGFSFileVersion();
    tgfsFileVersion.id = uuid();
    tgfsFileVersion.updatedAt = new Date();
    tgfsFileVersion.messageId = msg.messageId;
    tgfsFileVersion.size = msg.size.toJSNumber();
    return tgfsFileVersion;
  }

  static fromObject(
    tgfsFileVersionObject: TGFSFileVersionSerialized,
  ): TGFSFileVersion {
    const tgfsFileVersion = new TGFSFileVersion();
    tgfsFileVersion.id = tgfsFileVersionObject['id'];
    tgfsFileVersion.updatedAt = new Date(tgfsFileVersionObject['updatedAt']);
    tgfsFileVersion.messageId = tgfsFileVersionObject['messageId'];
    tgfsFileVersion.size =
      tgfsFileVersionObject['size'] ?? TGFSFileVersion.INVALID_FILE_SIZE;
    return tgfsFileVersion;
  }

  setInvalid() {
    this.messageId = TGFSFileVersion.EMPTY_FILE;
    this.size = TGFSFileVersion.INVALID_FILE_SIZE;
  }
}

export class TGFSFile {
  versions: { [key: string]: TGFSFileVersion } = {};
  latestVersionId: string;
  createdAt: Date;

  constructor(public readonly name: string) {}

  toObject(): TGFSFileObject {
    return {
      type: 'F',
      name: this.name,
      versions: this.getVersionsSorted().map((version) => version.toObject()),
    };
  }

  static fromObject(tgfsFileObject: TGFSFileObject): TGFSFile {
    const tgfsFile = new TGFSFile(tgfsFileObject.name);
    let lastUpdatedAt = 0;
    tgfsFile.createdAt = new Date();

    for (const version of tgfsFileObject.versions) {
      tgfsFile.addVersion(TGFSFileVersion.fromObject(version));
      if (version.updatedAt > lastUpdatedAt) {
        lastUpdatedAt = version.updatedAt;
        tgfsFile.latestVersionId = version.id;
      }
      if (version.updatedAt < tgfsFile.createdAt.getTime()) {
        tgfsFile.createdAt = new Date(version.updatedAt);
      }
    }

    return tgfsFile;
  }

  static empty(name: string): TGFSFile {
    const tgfsFile = new TGFSFile(name);
    tgfsFile.createdAt = new Date();
    tgfsFile.addEmptyVersion();
    return tgfsFile;
  }

  getLatest(): TGFSFileVersion {
    return this.getVersion(this.latestVersionId);
  }

  getVersion(uuid: string): TGFSFileVersion {
    return this.versions[uuid];
  }

  addVersion(version: TGFSFileVersion) {
    this.versions[version.id] = version;
  }

  addEmptyVersion() {
    const version = TGFSFileVersion.empty();
    this.addVersion(version);
    this.latestVersionId = version.id;
  }

  addVersionFromSentFileMessage(msg: SentFileMessage) {
    const version = TGFSFileVersion.fromSentFileMessage(msg);
    this.addVersion(version);
    this.latestVersionId = version.id;
  }

  updateVersion(version: TGFSFileVersion) {
    this.versions[version.id] = version;
  }

  getVersionsSorted(): TGFSFileVersion[] {
    return Object.values(this.versions).sort((a, b) => {
      return a.updatedAt.getTime() - b.updatedAt.getTime();
    });
  }

  deleteVersion(uuid: string) {
    delete this.versions[uuid];
  }

  isEmptyFile(): boolean {
    return this.getLatest().messageId === TGFSFileVersion.EMPTY_FILE;
  }
}

File: tgfs/model/file.py
import datetime
from typing import List, Iterable
from uuid import uuid4 as uuid
from dataclasses import dataclass, field

from tgfs.api.types import SentFileMessage
from .message import TGFSFileVersionSerialized, TGFSFileObject

EMPTY_FILE = -1
INVALID_FILE_SIZE = -1
INVALID_VERSION_ID = ""


@dataclass
class TGFSFileVersion:
    id: str
    updated_at: datetime.datetime
    message_id: int
    size: int

    @property
    def updated_at_timestamp(self) -> int:
        return int(self.updated_at.timestamp() * 1000)

    def to_dict(self) -> dict:
        return dict(
            type="FV",
            id=self.id,
            updatedAt=self.updated_at_timestamp,
            messageId=self.message_id,
            size=self.size,
        )

    @staticmethod
    def empty() -> "TGFSFileVersion":
        return TGFSFileVersion(
            id=str(uuid()),
            updated_at=datetime.datetime.now(),
            message_id=EMPTY_FILE,
            size=INVALID_FILE_SIZE,
        )

    @staticmethod
    def from_sent_file_message(message: SentFileMessage) -> "TGFSFileVersion":
        return TGFSFileVersion(
            id=str(uuid()),
            updated_at=datetime.datetime.now(),
            message_id=message.message_id,
            size=message.size,
        )

    @staticmethod
    def from_dict(data: TGFSFileVersionSerialized) -> "TGFSFileVersion":
        if (updated_at_ts := data.get("updatedAt", 0)) > 0:
            updated_at = datetime.datetime.fromtimestamp(updated_at_ts / 1000)
        else:
            updated_at = datetime.datetime.fromtimestamp(0)
        return TGFSFileVersion(
            id=data["id"],
            updated_at=updated_at,
            message_id=data["messageId"],
            size=data.get("size", INVALID_FILE_SIZE),
        )

    def set_invalid(self):
        self.message_id = EMPTY_FILE
        self.size = INVALID_FILE_SIZE


@dataclass
class TGFSFile:
    name: str
    latest_version_id: str = ""
    created_at: datetime.datetime = field(default_factory=datetime.datetime.now)
    versions: dict[str, TGFSFileVersion] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return dict(
            type="F",
            name=self.name,
            versions=[v.to_dict() for v in self.get_versions(sort=True)],
        )

    @staticmethod
    def from_dict(data: TGFSFileObject) -> "TGFSFile":
        versions = {
            v["id"]: TGFSFileVersion.from_dict(v) for v in data["versions"] if v
        }
        if versions:
            latest_version_id = max(
                versions, key=lambda k: versions[k].updated_at_timestamp
            )
        else:
            latest_version_id = INVALID_VERSION_ID
        return TGFSFile(
            versions=versions,
            latest_version_id=latest_version_id,
            created_at=datetime.datetime.now(),
            name=data["name"],
        )

    @staticmethod
    def empty(name: str) -> "TGFSFile":
        return TGFSFile(
            versions={},
            latest_version_id="",
            created_at=datetime.datetime.now(),
            name=name,
        )

    def get_latest_version(self) -> TGFSFileVersion:
        return (
            self.versions[self.latest_version_id]
            if self.latest_version_id
            else TGFSFileVersion.empty()
        )

    def get_version(self, version_id: str) -> TGFSFileVersion:
        return self.versions[version_id]

    def add_version(self, version: TGFSFileVersion) -> None:
        self.versions[version.id] = version
        if (
            self.latest_version_id == INVALID_VERSION_ID
            or version.updated_at > self.versions[self.latest_version_id].updated_at
        ):
            self.latest_version_id = version.id
        if not self.created_at or version.updated_at < self.created_at:
            self.created_at = version.updated_at

    def add_empty_version(self) -> None:
        version = TGFSFileVersion.empty()
        self.add_version(version)

    def add_version_from_sent_file_message(self, msg: SentFileMessage):
        version = TGFSFileVersion.from_sent_file_message(msg)
        self.add_version(version)
        return self.versions[self.latest_version_id]

    def update_version(self, version: TGFSFileVersion):
        self.versions[version.id] = version

    def get_versions(
        self, sort: bool = False, exclude_empty: bool = False
    ) -> List[TGFSFileVersion]:
        if not sort:
            res: Iterable[TGFSFileVersion] = self.versions.values()
        else:
            res = sorted(
                self.versions.values(),
                key=lambda v: v.updated_at_timestamp,
                reverse=True,
            )

        if exclude_empty:
            res = [
                v
                for v in res
                if v.message_id != EMPTY_FILE and v.size != INVALID_FILE_SIZE
            ]
        return list(res)

    def delete_version(self, version_id: str) -> None:
        if version_id not in self.versions:
            raise ValueError(f"Version {version_id} not found in file {self.name}.")
        del self.versions[version_id]
        if version_id == self.latest_version_id:
            if self.versions:
                self.latest_version_id = max(
                    self.versions, key=lambda k: self.versions[k].updated_at
                )
            else:
                self.latest_version_id = ""
